wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt

```diff
@@ -40,4 +52,221 @@ class Proto3WireProtocCompatibilityTests {
+    val jsonParser = JsonFormat.parser().usingTypeRegistry(typeRegistry)
+    val protocMessageDecodedFromWireJson = PizzaOuterClass.PizzaDelivery.newBuilder()
+        .apply { jsonParser.merge(moshiMessageJson, this) }
+        .build()
+
+    val wireMessageDecodedFromProtocJson = jsonAdapter.fromJson(protocMessageJson)
+
+    assertThat(protocMessageDecodedFromWireJson).isEqualTo(protocMessage)
+    assertThat(wireMessageDecodedFromProtocJson).isEqualTo(wireMessage)
+  }
+
+  @Test fun unregisteredTypeOnReading() {
+    val moshi = Moshi.Builder()
+        .add(WireJsonAdapterFactory())
+        .build()
+
+    val jsonAdapter = moshi.adapter(PizzaDelivery::class.java)
+
+    val json = """
+        |{
+        |  "address": "507 Cross Street",
+        |  "pizzas": [
+        |    {
+        |      "toppings": [
+        |        "pineapple",
+        |        "onion"
+        |      ]
+        |    }
+        |  ],
+        |  "promotion": {
+        |    "@type": "type.googleapis.com/squareup.proto3.pizza.BuyOneGetOnePromotion",
+        |    "coupon": "MAUI"
+        |  }
+        |}
+        """.trimMargin()
+
+    try {
+      jsonAdapter.fromJson(json)
+      fail()
+    } catch (expected: IOException) {
+      assertThat(expected).hasMessage("Cannot resolve type: " +
+          "type.googleapis.com/squareup.proto3.pizza.BuyOneGetOnePromotion")
+    }
+  }
+
+  @Test fun unrecognizedTypeOnWriting() {
+    val pizzaDelivery = PizzaDelivery(
+        address = "507 Cross Street",
+        pizzas = listOf(Pizza(toppings = listOf("pineapple", "onion"))),
+        promotion = AnyMessage.pack(BuyOneGetOnePromotion(coupon = "MAUI"))
+    )
+
+    val moshi = Moshi.Builder()
+        .add(WireJsonAdapterFactory())
+        .build()
+
+    val jsonAdapter = moshi.adapter(PizzaDelivery::class.java)
+    try {
+      jsonAdapter.toJson(pizzaDelivery)
+      fail()
+    } catch (expected: AssertionError) {
+      assertThat(expected).hasMessage("java.io.IOException: " +
+          "Cannot find type for url: " +
+          "type.googleapis.com/squareup.proto3.pizza.BuyOneGetOnePromotion")
+    }
+  }
+
+  @Test fun anyJsonWithoutType() {
+    val moshi = Moshi.Builder()
+        .add(WireJsonAdapterFactory())
+        .build()
+
+    val jsonAdapter = moshi.adapter(PizzaDelivery::class.java)
+
+    val json = """
+        |{
+        |  "address": "507 Cross Street",
+        |  "pizzas": [
+        |    {
+        |      "toppings": [
+        |        "pineapple",
+        |        "onion"
+        |      ]
+        |    }
+        |  ],
+        |  "promotion": {
+        |    "coupon": "MAUI"
+        |  }
+        |}
+        """.trimMargin()
+
+    try {
+      jsonAdapter.fromJson(json)
+      fail()
+    } catch (expected: JsonDataException) {
+      assertThat(expected).hasMessage(
+          """expected @type in JsonReader([text=\n    "coupon": "MAUI"\n  }\n}])""")
+    }
+  }
+
+  // Nice error on unrecognized @type
```

Comment:
Yes please

Reply:
This is actually what I thought I was already doing in `unregisteredTypeOnReading` and `unregisteredTypeOnWriting`.